src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+
+        let bits_per_block = self.operator.bits_per_block(ctxt);
+        let usize_bits: usize = mem::size_of::<usize>() * 8;
+
+        for (word_index, &word) in words.iter().enumerate() {
+            if word != 0 {
+                let base_index = word_index * usize_bits;
+                for offset in 0..usize_bits {
+                    let bit = 1 << offset;
+                    if (word & bit) != 0 {
+                        // NB: we round up the total number of bits
+                        // that we store in any given bit set so that
+                        // it is an even multiple of usize::BITS. This
+                        // means that there may be some stray bits at
+                        // the end that do not correspond to any
+                        // actual value; that's why we first check
+                        // that we are in range of bits_per_block.
+                        let bit_index = base_index + offset as usize;
+                        if bit_index >= bits_per_block {
+                            return;
+                        } else {
+                            f(bit_index);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn interpret_set<'c>(&self, ctxt: &'c O::Ctxt, words: &[usize]) -> Vec<&'c O::Bit> {
+        let mut v = Vec::new();
+        self.each_bit(ctxt, words, |i| {
+            v.push(self.operator.interpret(ctxt, i));
+        });
+        v
+    }
+}
+
+pub trait BitwiseOperator {
+    /// Joins two predecessor bits together, typically either `|` or `&`
+    fn join(&self, pred1: usize, pred2: usize) -> usize;
+}
+
+/// Parameterization for the precise form of data flow that is used.
+pub trait DataflowOperator : BitwiseOperator {
+    /// Specifies the initial value for each bit in the `on_entry` set
+    fn initial_value() -> bool;
+}
+
+pub trait BitDenotation: DataflowOperator {
+    /// Specifies what is represented by each bit in the dataflow bitvector.
+    type Bit;
+
+    /// Specifies what, if any, separate context needs to be supplied for methods below.
+    type Ctxt;
+
+    /// A name describing the dataflow analysis that this
+    /// BitDenotation is supporting.  The name should be something
+    /// suitable for plugging in as part of a filename e.g. avoid
+    /// space-characters or other things that tend to look bad on a
+    /// file system, like slashes or periods. It is also better for
+    /// the name to be reasonably short, again because it will be
+    /// plugged into a filename.
+    fn name() -> &'static str;
+
+    /// Size of each bitvector allocated for each block in the analysis.
+    fn bits_per_block(&self, &Self::Ctxt) -> usize;
+
+    /// Provides the meaning of each entry in the dataflow bitvector.
+    /// (Mostly intended for use for better debug instrumentation.)
+    fn interpret<'a>(&self, &'a Self::Ctxt, idx: usize) -> &'a Self::Bit;
```

Comment:
I'm +1 for that if you can write a good API for that. The current `interpret` API is useful only for graphviz.